### ISSUE TITLE
React 15.0.1 Updates causing issues with DOM

### ISSUE THREAD

**Issue by nickbae91** [facebook/react#6472]
Hey,
So I'm a fairly new developer who had a working app until version v0.14.8, and just updated to v15.0.1 today. The update did remove a few errors/warnings I was struggling with (so a great relief there), but brought up this issue.

Specifically, the app itself renders fine, with all the pages looking fine (all components, text, etc. in the right places), but is just "frozen" - completely non-functional, not allowing me to click on anything etc.

The error message I see are as follows (from Chrome and Firefox):
![capture](https://cloud.githubusercontent.com/assets/12439968/14408726/7d6d4a18-febd-11e5-8d84-fb1633d1cc08.PNG)
![capture2](https://cloud.githubusercontent.com/assets/12439968/14408728/80ddd280-febd-11e5-9397-67bcb0f03f30.PNG)

I back-tracked it to where I could and found this being the part where the "node" is coming off as null:
![capture3](https://cloud.githubusercontent.com/assets/12439968/14408739/cf942208-febd-11e5-898a-75e25f008695.PNG)

So I have absolutely no clue what's going on and there seems to be no hint as to what exactly is causing this issue. I will try ripping the app apart and do a thorough debugging, but wouldn't expect much from it (due to my unfamiliarity with development yet...).
I really really appreciate the help in advance, and hope that this problem can contribute somewhat to improving the React ecosystem!

**Comment by roganov** [comment]
React 0.14 warns about deprecations that are removed/changed in React 15. I would suggest to resolve all the errors/warning in 0.14 and then upgrade.

**Comment by gaearon** [comment]
Please make sure you don’t have two instances of React on the same page.
`npm ls react` can help to determine this.

If you don’t, please share the minimal example reproducing this issue on GitHub or JSFiddle.
Thanks!

**Comment by doctyper** [comment]
I am also seeing this error, particularly in unit tests that do not unmount components after the tests complete. In these cases, the value of [`rootNode.parentNode`](https://github.com/facebook/react/blob/master/src%2Frenderers%2Fdom%2Fclient%2FReactEventListener.js#L36) is `null`, presumably because the node is no longer attached to the DOM. The null value is then passed into [`getClosestInstanceFromNode`](https://github.com/facebook/react/blob/master/src%2Frenderers%2Fdom%2Fclient%2FReactDOMComponentTree.js#L110-L138), where the error is thrown due to `node` being null.

My current workaround is to return null if the node is null:

```
if (node == null) {
    return null;
}
```

Which seems to work, though I admittedly don't know the full consequences for doing so. I think this is at least a candidate for an invariant warning/error.

**Comment by doctyper** [comment]
I should add, this error surfaces after testing with [`ReactTestUtils.renderIntoDocument`](https://github.com/facebook/react/blob/master/src%2Ftest%2FReactTestUtils.js#L79-L87), which:

1. does not have a way to unmount a test component
2. does not attach the test component to the document

I believe this would explain why [`parentNode` would be null](https://developer.mozilla.org/en-US/docs/Web/API/Node/parentNode).

> It also returns null if the node has just been created and is not yet attached to the tree.

**Comment by jimfb** [comment]
@nickbae91 @doctyper I am still unable to reproduce.  Can one of you provide a jsfiddle that demonstrates the issue?f

**Comment by vinnymac** [comment]
I actually debugged this very issue yesterday morning while trying to upgrade to v15. The solution I came up with was to provide ReactDOM with an element that already existed in the DOM via `document.getElementById`.

One of the projects I work on has an old backbone codebase and it has slowly been migrated to react. The element that ReactDOM was being told to render to was a div that backbone generated for me. Previously (before v15) this element was never being inserted into the actual DOM.

My stack trace revealed that `getClosestInstanceFromNode` had a null node as @doctyper says above. Debugging this further I discovered that `ReactEventListener.js` was the one calling `getClosestInstanceFromNode`. If you take a look at [findParent](https://github.com/facebook/react/blob/master/src/renderers/dom/client/ReactEventListener.js#L28) you can see where the node gets defined. In my case the rootNode was an audio element that a react class was producing. The audio element had no parentNode, and so container became null.

While debugging this issue occasionally it would just disappear, and stop reporting. I would refresh a dozen times and see it again.

I reproduced the behavior I describe above in a fiddle. https://jsfiddle.net/mvsgkg71/

I couldn't get it to reproduce at first. After clicking run ten times consecutively it finally triggered, here is a screenshot proving it actually happens.

![Example](http://i.imgur.com/dckiU4Y.png)

**Comment by jimfb** [comment]
Ok, this is starting to look like a Chrome bug to me (race condition in chrome, potentially related to garbage collection).  Has anyone observed this behavior in a browser other than Chrome?

I have a (relatively) consistent repro: https://jsfiddle.net/r4ewebqs/

**Comment by vinnymac** [comment]
@jimfb The OP has a screenshot of the error appearing in firefox. I attempted your fiddle as well as my own in Safari and Firefox, and did not get any errors.

@nickbae91 could you provide additional information, such as the version of the browsers you tested with?

**Comment by jimfb** [comment]
Yeah, if I upgrade from Chrome 49 (out-of-date) to Chrome 50 (latest stable), the error goes away.  This is consistent with my original observation that it appeared to be a race condition potentially related to garbage collection.

I'm going to close this out as a browser bug, unless someone is able to reproduce on Chrome 50.  I'd also be curious to see a firefox repro.

**Comment by zpao** [comment]
Let's be careful about assuming a GC issue / browser bug. We should be able to confidently say what specifically is at play. It could well be a bug in React (particularly around transactions, mounting, and local events)

My best guess here (especially since all the reports are for nodes that we have to add load / error listeners to directly) is that timing is at play. With the switch to createElement we'll actually kick off the fetching of resources earlier than before - instead of waiting for the browser to parse the img element and see src, we create the img and set src right there. That means we'll probably get a load event earlier. How early and is it possible for that event to fire before we actually attach a listener to that node is the question, and what impact does it actually have. Does it only happen on updates? Are there other events actually at play?

> One thing I find interesting is if you edit the fiddle so that the audio element has no source elements, the errors do not appear

This would support my theory here as you won't get at least some of the media events.

**Comment by vinnymac** [comment]
@jimfb I can confirm. Upgraded to Chrome 50 and I am no longer able to reproduce the issue. Hoping we can get some other reproductions of this from the OP. @nickbae91 was the only one to report seeing this elsewhere. In the meantime I'll think about a way to support @zpao's theory.

**Comment by jmullo** [comment]
I also bumped into something similar with unit tests running in PhantomJS 2.1.1. Worked fine in 0.14.8:

```
undefined is not an object (evaluating 'node[internalInstanceKey]')
getClosestInstanceFromNode@...
getInstanceFromNode@...
```

**Comment by gaearon** [comment]
Please try to provide a minimal example reproducing this.

**Comment by jmullo** [comment]
Seems to be some kind of timing issue, as the cases pass if I delay calls to ReactTestUtils.Simulate.focus() or ReactTestUtils.Simulate.click().

**Comment by haibo3318** [comment]
I also saw this error when rendering <img />. But when I remove the "src" attribute in <img />, this error disappeared.  
![image](https://cloud.githubusercontent.com/assets/15971704/15733758/68d1be0a-28dd-11e6-9cea-f3d93f24c0fb.png)

This error only exist when using npm package with webpack require. Everything is OK with individual build.

**Comment by torarnek** [comment]
I can confirm that several of our mobile users on Android, using Chrome 34.0.1847, are getting this error. We are on React 15.1.0.

**Comment by jimfb** [comment]
@torarnek There is very little we can do without a repro.  If you can share the code where this is occurring, we can dig deeper.  Otherwise, this is inactionable.

haibo3318 mentioned that things appeared to work with the individual builds / bower builds.  Do you find the same to be true?

**Comment by gauravChaturvedi** [comment]
I'm facing the same issue, albeit for me it is occurring due to a video tag.
<img width="633" alt="screen shot 2016-06-27 at 10 12 59 pm" src="https://cloud.githubusercontent.com/assets/8039449/16394227/5ebf3d76-3cb4-11e6-8fa4-f91edfa4477d.png">

**Comment by gauravChaturvedi** [comment]
I was able to solve this by doing the following: 

Instead of declaring my parent HTML Node inside index.html as :- 

<body>
    <div id="root"></div>
    <script src="./bundle.js" type="text/javascript"></script>
</body>


I removed the <div> from index.html and added the div from my js code as follows :- 

const root = document.createElement('div');
document.body.appendChild(root);
ReactDOM.render(<App />, root);


Hope this works for others as well !

**Comment by gaearon** [comment]
@gauravChaturvedi Can you share a minimal project that reproduces the problem?

**Comment by torarnek** [comment]
This problem was introduced by using Redux connect on various sub-components, and since this was a critical production feature, we did a quick refactoring to only connect the main component, and now it seems to be working fine.   

The problem seems affect media tags when doing batch updates. See my previous stacktrace.

**Comment by gaearon** [comment]
Unfortunately until somebody can share a reproducing example we will be left guessing as to what it is. It would really help if you could spare a few hours to find a minimal reproducing sample. Hopefully React saved you more than a few hours so that seems fair :-)

**Comment by torarnek** [comment]
Hey Dan, if I knew how to create a reproducible example I would have definitely made one (you guys deserve it), however in our case only android mobile users with a certain Chrome browser got this issue.  I tried to reproduce with the android emulator, but was not able to get the error. So I am sorry about that.

**Comment by gaearon** [comment]
This is very helpful info, thanks for sharing! Maybe somebody else will be able to use it to publish a case reproducing on Androids.

**Comment by vinnymac** [comment]
##### Summary
I stopped seeing it in `Chrome 50+`, and have not been able to reproduce the issue since. I use sentry to track errors in several of my projects. I can see the same audio element issue above occurring for certain users, and have compared the stack traces.

##### Insight
These users were running `Opera 38.0.2220.31` and `IE 11.0`. They were using `Windows 7` at the time. They were all running `v15.1.0` of React. Several months of stack traces and no other browser has reported this error. I don't have many mobile web users (native apps), so I wouldn't expect to see those browsers reported (<1%).

##### Example Stack Trace
```javascript
TypeError: Cannot read property '__reactInternalInstance$1yks451snh16k09eiy17cik9' of null
  at Object.s [as getClosestInstanceFromNode](/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:123:18949)
  at r(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:139:26220)
  at i(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:139:26426)
  at r.perform(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:124:12191)
  at Object.batchedUpdates(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:139:26041)
  at Object.i [as batchedUpdates](/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:124:6959)
  at dispatchEvent(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:139:27375)
  at HTMLAudioElement.r(/assets/application-eebe52b07353e35ecb012c38735fcf9c.js:3:9163)
```

##### Reproduction Attempt
I spun up a VM of `Windows 7` with the same versions of `IE` and `Opera` above, and ran our fiddles and attempted to reproduce to no avail. If someone wants to give this a shot on their own, you can download [Opera here](http://www.opera.com/computer) and a copy of [Windows 7 with IE 11 here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/). Hopefully this new information will lead someone else to a reproduction. If anyone is keeping track of bugs and is seeing a similar browser pattern that may be useful to learn.

**Comment by johndugan** [comment]
Hi Folks,

I ran into this error too. I am using the `html-webpack-plugin`. I had a call for my js file in an `index.template.html` file I defined for the `template` property. Because the html webpack plugin injects the JavaScript bundle, my js file was being called twice. You can pass `inject: false` to the plugin's configuration object to prevent this behavior, or remove the `<script>` call in your template file. This will solve the error resulting from the `getClosestInstanceFromNode` function (if it stems from the plugin).
```
    plugins: [
        new webpack.optimize.OccurenceOrderPlugin(), // https://github.com/webpack/docs/wiki/optimization
        new webpack.HotModuleReplacementPlugin(),
        new webpack.NoErrorsPlugin(),
        new HtmlWebpackPlugin({
            template: 'index.template.html',
            inject: false,
            minify: {
                collapseWhitespace: true
            }
        })
    ]
```

Below are some screenshots to clarify.

_Note the explicit_ `<script>` _tag in my `index.template.html` file._
![ice_screenshot_20160902-135916](https://cloud.githubusercontent.com/assets/1389303/18220560/e1ea98a4-7138-11e6-8592-039b8e61d693.png)

_Note that I am accepting the default_ `inject: true` _behavior of the html webpack plugin._
![ice_screenshot_20160902-135844](https://cloud.githubusercontent.com/assets/1389303/18220596/4e130ffc-7139-11e6-8ceb-e4e5a1352b7a.png)

_Note the two_ `<script>` _calls._
![ice_screenshot_20160902-135801](https://cloud.githubusercontent.com/assets/1389303/18220615/7c11c2c2-7139-11e6-9a81-b648633a5eac.png)

**Comment by joshunger** [comment]
React: 15.3.2
Browsers:  Google Chrome 56.0.2913.3 and FireFox 52.0a1 (2016-11-02) (64-bit)

```
Uncaught TypeError: Cannot read property '__reactInternalInstance$795sokf19kpfem9nusdbc323xr' of null
    at Object.getClosestInstanceFromNode (ReactDOMComponentTree.js:107)
    at findParent (ReactEventListener.js:39)
    at handleTopLevelImpl (ReactEventListener.js:68)
    at ReactDefaultBatchingStrategyTransaction.perform (Transaction.js:138)
    at Object.batchedUpdates (ReactDefaultBatchingStrategy.js:63)
    at Object.batchedUpdates (ReactUpdates.js:98)
    at dispatchEvent (ReactEventListener.js:150)
    at HTMLImageElement.a.bugsnag (bugsnag-3.min.js:1)
```

![image](https://cloud.githubusercontent.com/assets/2301847/20180023/dfa9f086-a715-11e6-94c1-4b1103d61fed.png)

@gaearon @jimfb repros on my machine for the next hour if you want me to debug into it let me know specific breakpoints

**Comment by nikhilgoud** [comment]
React:15.3.0 , webpack:1.13.1
I am fairly new to react, working on a common dialog modal, where request(ajax) response needs to be displayed, 
Created a component dialog-modal(used bootstrap modal)
![image](https://cloud.githubusercontent.com/assets/21196948/20217675/eef23dbe-a846-11e6-828b-30afec0fbabc.png)

![image](https://cloud.githubusercontent.com/assets/21196948/20217484/21e349ee-a846-11e6-9f18-077ad62e48f8.png)

and I was able to get the error
![image](https://cloud.githubusercontent.com/assets/21196948/20217735/3302797e-a847-11e6-9189-1f946b3d52ab.png)

All of this happened because I was removing the modal when onHide line number 8 in dialog-modal component, on changing that to just hide the error was gone. :)
Before actually committing the transaction I deleted the node from DOM.
Hope this helps some one.

**Comment by osbertngok** [comment]
I got the same error due to exactly the reason @johndugan brought up.

**Comment by sdeleon28** [comment]
This issue is happening to me in Google Console's Fetch as Google (Fetch and Render). Works fine in Chrome. If I add a snippet to display errors in the DOM, Fetch as Google shows me this:

![fetch_as_google_error](https://cloud.githubusercontent.com/assets/633057/22622440/b82ebf4a-eb18-11e6-97df-ef44bedaa211.jpg)

I have managed to narrow this down to a few lines of code. If I remove these lines and bundle up my app with Webpack again, then the error disappears and Google is able to render the page just fine:

```javascript
const images = this.props.images.map((image, index) => ({
  ...image,
  index,
  selected: index === this.state.selectedImageIndex,
}));
```

The snippet seems like a totally random piece of ES6, so I'm not sure what's going on here. This doesn't look 100% related to the problem other users are facing but the stack trace looks very similar.

**Comment by sdeleon28** [comment]
This happened to me because I had an unhandled exception in the `render` method of a component. The error only happened in Google Console's Fetch as Google. I didn't have the appropriate polyfill for the Googlebot agent to run my code properly. In this particular case, the offending line called `Array.prototype.find` . Changing `myArray.find(myCondition)` to `myArray.filter(myCondition)[0]` solved the problem.

**Comment by thomas-siimpl** [comment]
The following would result in the `Uncaught TypeError: Cannot read property '__reactInternalInstance$gf9gtshwv1avu0220061ug14i' of null` whenever navigating.

```
  render(
      <Provider store={store}>
        <App>
          <Layout user={window.env.user}>
            {components}
          </Layout>
        </App>
      </Provider>,
    document.getElementById('client')
  )
```

the remedy was to wrap things in one more element, new root level `<div />`.

```
  render(
    <div>
      <Provider store={store}>
        <App>
          <Layout user={window.env.user}>
            {components}
          </Layout>
        </App>
      </Provider>
    </div>,
    document.getElementById('client')
  )
```

**Comment by destromas1** [comment]
@gaearon 
We are using React 15.4.0 and ReactDOM 15.4.0 with TypeScript , Webpack
Then we are transpiling to ES5 for IE11 support.

In Chrome, Edge our app works fine.
But getting error in IE11

```
SCRIPT5007: Unable to get property '@@__symbol:iterator0.57059978612322841' of undefined or null reference
polyfill.min.js (7,19646)

SCRIPT5007: Unable to get property '__reactInternalInstance$5hya2doizuj5' of undefined or null reference
react-dom.min.js (13,23434)
```

**Comment by destromas1** [comment]
BTW , it was fixed after changing Polyfill CDN to `babel-polyfill` with Webpack Entry

**Comment by typify** [comment]
I had a similar issue. It was resolved as soon as I closed the Firebug window. I couldn't get to the root cause, but at least this might help others as well. Took me 2 hours to find out..

**Comment by aspiers** [comment]
I can confirm @joshunger's [observation](https://github.com/facebook/react/issues/6472#issuecomment-259701256) that this happens when there is an uncaught exception in `mapStateToProps`.  I see the same thing with React 15.4.2 and Chrome 57.0.2987.110 (64-bit), i.e. the console shows this trace three times in a row:

```
ReactDOMComponentTree.js:113 Uncaught TypeError: Cannot read property '__reactInternalInstance$x760fmy243' of null
    at Object.getClosestInstanceFromNode (ReactDOMComponentTree.js:113)
    at findParent (ReactEventListener.js:38)
    at handleTopLevelImpl (ReactEventListener.js:67)
    at ReactDefaultBatchingStrategyTransaction.perform (Transaction.js:140)
    at Object.batchedUpdates (ReactDefaultBatchingStrategy.js:62)
    at Object.batchedUpdates (ReactUpdates.js:97)
    at dispatchEvent (ReactEventListener.js:147)
```

Luckily it did then display the error caused by the bug in `mapStateToProps`, so I was able to locate the bug and fix it.

**Comment by lyhoanglong** [comment]
I got this error too, but dont know why this happen. This is my code
```
import React from 'react';
import Logo from '../styles/images/logo.svg';
import '../styles/styles.scss';

const App = props => (
  <div className="wrapper">
    <div className="beleaf-logo">
      <img src={Logo} alt="Beleaf" />
      <p>Beleaf</p>
    </div>
  </div>
);

export default App;
```
If I removed image everything ok. Anyone know what happen here?

**Comment by destromas1** [comment]
@lyhoanglong 
Why dont you use image directly inside `src="../styles/images/logo.svg"` something like this ?

**Comment by lyhoanglong** [comment]
@destromas1 because sometime I can reuse this image. What's wrong if I use image this way?

**Comment by ugene1213** [comment]
@lyhoanglong did you resolve your issue? I'm getting the same exact error as you and it appears every time I try to render an image

**Comment by lyhoanglong** [comment]
@ugene1213 I didnt and dont know how to fix this.

**Comment by divyanshupathak** [comment]
Hi @lyhoanglong 
remove `import Logo from '../styles/images/logo.svg';` and try to pass the `src` of image or `filecursor` by using `props` from its parent component.

**Comment by lasfin** [comment]
Got the same error (when navigate to another page of app by clicking to link). Can fix it by changing root node className, but it isn't possible in my case.

**Comment by dervalp** [comment]
Hi,

Got the same kind of error with an img element.

> ReactDOMComponentTree.js:113 Uncaught TypeError: Cannot read property '__reactInternalInstance$c12uup2dzwt' of null

```
import React from 'react';
import LogoImage from './Logo.png';


const Logo = () => {
  return (
    <div>
      <img src={LogoImage} alt="Logo"/>
    </div>
  )
};
```
I am using url-loader but even if I am hardcoding the value of the src attribute, I got the error.

```
"react": "^15.5.4",
"react-dom": "^15.5.4",
```

```
"babel-cli": "^6.24.1",
"babel-core": "^6.24.1",
"babel-jest": "^11.0.1",
"babel-loader": "^7.0.0",
"file-loader": "^0.11.1",
"html-webpack-plugin": "^2.28.0",
"prop-types": "^15.5.8",
"url-loader": "^0.5.8",
"webpack": "^2.5.1",
"webpack-dev-server": "^2.4.5"
```

**Comment by alexandernanberg** [comment]
I hade the exact same issue, the problem for me was that I included the javascript files multiple times as @johndugan pointed out. Would it be possible for React to throw a warning if it senses that it's included multiple times?

**Comment by locomotif** [comment]
I am having the same issue around the `<img src=".." />` tag on a SPA.  
@gaearon I have created a minimal project to reproduce the issue.

Thanks!

**Required File structure**
```txt
|-- .babelrc
|-- app
    |-- app.jsx
    |-- index.ejs
|-- package.json
|-- public
|-- webpack.config.js
```
**app.jsx**
```js
// app.jsx
import React, { Component } from 'react';
import ReactDOM from 'react-dom';

class App extends Component {
    render() {
        return (
            <img src="https://facebook.github.io/react/img/logo.svg" />
        );
    }
}

ReactDOM.render(
    <App/> ,
    document.getElementById('container')
);
```
**webpack.conf.js**
```js
// webpack.conf.js
const path = require('path');
const webpack = require('webpack');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');

module.exports = {
    context: path.resolve(__dirname, 'app'),
    entry: {
        app: './app.jsx'
    },
    output: {
        filename: 'js/[name].js',
        chunkFilename: 'js/[id].js',
        path: path.resolve(__dirname, 'public')
    },
    module: {
        rules: [
            {
                test: /\.(css|scss)$/,
                use: ExtractTextPlugin.extract({
                    use: [
                        { loader: 'css-loader' },
                        { loader: 'sass-loader' }
                    ]
                })
            },
            {
                test: /\.(js|jsx)$/,
                use: [{ loader: 'babel-loader', }],
                exclude: /node_modules/
            },
            {
                test: /\.(png|jpg|svg)$/,
                use: [{ loader: 'file-loader?name=images/[name].[ext]' }]
            }
        ]
    },
    plugins: [
        new ExtractTextPlugin('css/[name].css'),
        new webpack.HotModuleReplacementPlugin(),
        new HtmlWebpackPlugin({
            template: 'index.ejs'
        })
    ]
};
```
**.babelrc**
```json
{
    "presets":[ 
        "es2015",
        "react"
    ],
    "plugins":[
        "transform-object-rest-spread"
    ]
}  
```
**package.json**
```json
{
  "name": "example",
  "version": "0.0.1",
  "description": "img src issue",
  "scripts": {
    "start": "webpack-dev-server",
    "build": "webpack"
  },
  "devDependencies": {
    "babel-core": "^6.25.0",
    "babel-loader": "^7.0.0",
    "babel-plugin-transform-object-rest-spread": "^6.23.0",
    "babel-preset-es2015": "^6.24.1",
    "babel-preset-react": "^6.24.1",
    "css-loader": "^0.28.4",
    "extract-text-webpack-plugin": "^2.1.2",
    "file-loader": "^0.11.2",
    "html-webpack-plugin": "^2.28.0",
    "react": "^15.6.1",
    "react-dom": "^15.6.1",
    "scss-loader": "^0.0.1",
    "webpack": "^2.6.1",
    "webpack-dev-server": "^2.4.5"
  },
  "dependencies": {}
}
```
**index.ejs**
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>Img src issue</title>
  </head>
  <body>
    <div id="container"></div>
    <script src="/js/app.js"></script>
  </body>
</html>
```

**Comment by locomotif** [comment]
Also, I confirmed @doctyper findings that `rootNode.parentNode` is `null`.  I did a similar workaround in *react-dom/lib/ReactEventListner.js*, function *findParent()*.  I added the following, just so that I can bypass the error:

```
  if (container) {
      return ReactDOMComponentTree.getClosestInstanceFromNode(container);
  } else {
      return null;
  }
```

I am just beginning to learn react, so I am not sure exactly what is happening and why there is a detached node looking for a parent.

**Comment by locomotif** [comment]
Also, I confirmed @doctyper findings that rootNode.parentNode is null. I did a similar workaround in react-dom/lib/ReactEventListner.js, function findParent(). I added the following, just so that I can bypass the error:

  if (container) {
      return ReactDOMComponentTree.getClosestInstanceFromNode(container);
  } else {
      return null;
  }
I am just beginning to learn react, so I am not sure exactly what is happening and why there is a detached node looking for a parent.

**Comment by cdaringe** [comment]
I haven't pin-pointed it, but i'm with @locomotif--for me the single change that either creates or eliminates the failure is the `src` attribute on the image tag.

it's not a minimum viable example, but it is _rather small_--almost === CRA barebones. if one:

- [clones this snapshot](https://github.com/dino-dna/truth.lol/tree/ccdc21b3a901f176e0cf8af205a7a35ebc4d58c1/packages/ui)
- runs `yarn && yarn start`, you can see the error
- fiddle with the `<App />` component's image to see it get weird.

**Comment by cdaringe** [comment]
it appears that `rootNode` isn't guaranteed to be mounted on the DOM after calling `getNodeFromInstance` from within `findParent`.

consequently, sniffing for `rootNode`'s `.parentNode` yields `null`.

<img width="837" alt="screen shot 2017-06-18 at 8 15 24 pm" src="https://user-images.githubusercontent.com/1003261/27268205-f972c092-5462-11e7-9e7d-187f4a947433.png">

missing from above, `rootNode.id === 'id'`

**Comment by alexandernanberg** [comment]
@locomotif You are having the same problem that I was dealing with. Just remove the <script> tag in your index.ejs file and it should work as expected. The issue is that you are including your script files 2 times. You could also set HtmlWebpackPlugin to not automatically include your assets if you want to do that in your template.

**Comment by locomotif** [comment]
@alexandernanberg That makes absolute sense! thanks, that did the trick!  I had just recently added the `html-webpack-plugin`, and it didn't occur to me to remove that line of code.  I had noticed that my console message was appearing twice, and briefly assumed that something was being loaded twice; but then I disregarded it and thought it had to do with the hot loader.  Not until I added the image node a couple of days after did this issue exposed itself.  Thanks a million, I don't think I would have thought to look there since the **index.ejs** was not in my radar.

**Comment by jeantimex** [comment]
I was experiencing the same issue as @haibo3318 did. At first glance, the src attribute looks suspicious, because everything works fine after removing the src attribute from the img tag. Then I took another look into the html content, I found this:
```
<script src="main.js" ... />
<script src="main.js" ... />
```
In my case, I have two little separate React widgets on the page, and they are embedded by the same php template in which the script tag is included. This caused the `Uncaught TypeError: Cannot read property '__reactInternalInstance$...` error!

The way I fixed this problem is just making sure the script tag is included only once.

**Comment by soundobj** [comment]
This is still happening on Chrome Version 60.0.3112.90 (Official Build) (64-bit) and react 15.6.1 on Windows.

Uncaught TypeError: Cannot read property '__reactInternalInstance$

On my particular example the culprit are <img> tags. if i remove the src attribute on the images then  the error goes away. Firefox is fine on the same machine as OSX / Chrome so it seems a browser / OS specific issue.

**Comment by gaearon** [comment]
If you can reproduce this please attach a reproducible example. Just saying "I have this too" is not very helpful because it does not help us understand the problem, and does not guarantee you have excluded known causes (such as duplicate React).

**Comment by colindekker** [comment]
Have this issue now, here is a bare bones SPfx project, including the React draftjs wysiwyg component. Works fine on chrome, IE11 and Edge the load fails. run `gulp serve` to load the SPfx workbench, after it loads, click the little plus at the top of the page and select RteEditor to add the component in the workbench and it will fail. Project: https://drive.google.com/file/d/0BxopWiNF_Y9zWmFPYk5IaEFZNms/view?usp=sharing

**Comment by WGLayner** [comment]
I met the same problem with below code. The problem is not there after remove img tag.

<Navbar.Header>
          <Navbar.Brand>
            <a href="#">
              <img src="/images/cp1.png" alt="logo" style={logoStyle} />
            </a>
          </Navbar.Brand>
          <Navbar.Toggle />
        </Navbar.Header>

**Comment by colindekker** [comment]
I replaced it with font-awesome icons

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by gaearon** [comment]
In the vast majority of cases this error was caught by an earlier error in the application code that got swallowed by some third-party library (or application code).

React 16 should surface the original error better in this case so I’m closing. If you still experience this after upgrading to React 16 please file a new issue with a complete reproducing example.

**Comment by tkssharma** [comment]
one possible solution is you just need to make sure if you have multiple instances loaded in you index.html of react. I was getting it because i have added webpack bundle twice while running it using webpack-dev-server with htmlwebpack plugin. 

now fixed !! :)

**Comment by chaituckr** [comment]
This issue is also occurs when we dont import the child/third party components and use them in the current component. So, make sure you imported all the components that are being used in current component.

**Comment by cancerberoSgx** [comment]
this happens to me for some reason too, probably because my component is being updated multiple times. I have to do this before calling `simulate('click')`

```
wrapper.filterWhere(w=>w && w.getDOMNode()) .forEach(w=>
    w.simulate('click', ...
```

w.parent() is not enough. I think w.exists() here should return false or simply add this checking in simulate() implementation.
